File: Strings/SortCharByFrequency.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
using namespace std;
class Solution{
  private:
    static bool comparator(pair<int,char> p1,pair<int,char> p2){
      if(p1.first<p2.first){
        return false;
      }
      if(p1.first>p2.first){
        return true;
      }
      return p1.second<p2.second;
    }
  public: 
    vector<char> freqSort(string s){
      pair<int,char> freq[26];
      for(int i=0;i<26;i++){
        freq[i]={0,(char)'a'+i};
      }
      for(char ch:s){
        freq[ch-'a'].first++;
      }
      sort(freq,freq+26,comparator);
      vector<char> ans;
      for(int i=0;i<26;i++){
        if(freq[i].first>0){
          ans.push_back(freq[i].second);
        }
      }
      return ans;
    }
};
int main(){
  Solution sol;
  string s="trestzzz";
  vector<char> res=sol.freqSort(s);
  for(char c:res){
    cout << c << " ";
  }
  return 0;
}

File: Strings/SortCharByFrequency.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
class Solution{
  public List<Character> frequencySort(String s){
    Pair[] freq=new Pair[26];
    for(int i=0;i<26;i++){
      freq[i]=new Pair(0,(char)('a'+i));
    }
    for(char ch:s.toCharArray()){
      freq[ch-'a'].freq++;
    }
    Arrays.sort(freq,(p1,p2)->{
      if(p1.freq!=p2.freq){
      return p2.freq-p1.freq;//descending
      }
      return p1.ch-p2.ch;//ascending
    });
   List<Character> res=new ArrayList<>();
   for(Pair p:freq){
     if(p.freq>0){
       res.add(p.ch);
     }
   }
   return res;
  }
  class Pair{
    int freq;
    char ch;
    Pair(int freq,char ch){
      this.freq=freq;
      this.ch=ch;
    }
  }
}
public class SortCharByFrequency{
  public static void main(String[] args){
    Solution sol=new Solution();
    String s="tree";
    List<Character> result=sol.frequencySort(s);
    System.out.println(result);
  }
}